applicationset/generators/git.go

```diff
@@ -238,8 +261,33 @@ func (g *GitGenerator) generateParamsFromGitFile(filePath string, fileContent []
 	return res, nil
 }
 
+// filterFiles filters a list of file paths based on inclusion and exclusion patterns defined in GitFileGeneratorItems.
+// It uses doublestar globbing to support complex patterns like "**/*.yaml".
+func (g *GitGenerator) filterFiles(files []argoprojiov1alpha1.GitFileGeneratorItem, allFilePaths []string) ([]string, error) {
+	var res []string
+	for _, filePath := range allFilePaths {
+		appInclude := false
+		for _, file := range files {
+			match, err := doublestar.Match(file.Path, filePath)
+			if err != nil {
+				return nil, fmt.Errorf("error while matching file path %q with pattern %q: %w", filePath, file.Path, err)
+			}
+			if match {
+				appInclude = !file.Exclude
+			}
+		}
+		// append only those file paths in the result that satisfies the path pattern
+		if appInclude {
+			res = append(res, filePath)
+		}
+	}
+	return res, nil
+}
```

Comment:
The exclusions should get the results from `g.repos.GetFiles` and remove the results from the list of included files list if they exist. It should not evaluate the existing result based on `doublestar.Match`.

Reply:
So I tried multiple ways to see how can we implement it. First, we are trying to range over all the [file generator items](https://github.com/argoproj/argo-cd/pull/22734/files#diff-e18bda08876f1471ded096f197d58778d250cd942ca66539391b067d5f069f79R149) , so let's say that a file matches my first pattern with exclude set to false. Now, the same file matches my other pattern with exclude set to true, this file will be included with the current codebase.  What's the best way via which this logic can be implemented on the server side?
cc @rumstead / @crenshaw-dev

Reply:
This way the globbing pattern is respected as provided by the user.

Reply:
It worked but are we okay with this approach?
![image](https://github.com/user-attachments/assets/bff0aa33-8ead-40c0-91bc-0d6a8c4de25f)

Reply:
Testing this is a bit complex.

Reply:
This makes sense to me! It's a bummer to make multiple `LsFiles` calls. Maybe we should parameterize that on the repo-server side so we make fewer calls. But that can be a future enhancement.